stablehlo/reference/Element.cpp

```diff
@@ -120,6 +120,114 @@ Element Element::operator+(const Element &other) const {
   report_fatal_error(std::move(err));
 }
 
+Element max(const Element &e1, const Element &e2) {
+  auto type = e1.getType();
+
+  if (isSupportedFloatType(type)) {
+    auto left = getFloatValue(e1);
+    auto right = getFloatValue(e2);
+    return Element(type, FloatAttr::get(type, maximum(left, right)));
+  }
+
+  if (isSupportedUnsignedIntegerType(type)) {
+    auto left = getIntegerValue(e1);
+    auto right = getIntegerValue(e2);
+    return left.ugt(right) ? e1 : e2;
+  }
+
+  if (isSupportedSignedIntegerType(type)) {
+    auto left = getIntegerValue(e1);
+    auto right = getIntegerValue(e2);
+    return left.sgt(right) ? e1 : e2;
+  }
+
+  if (isSupportedComplexType(type)) {
+    auto leftComplexValue = getComplexValue(e1);
+    auto rightComplexValue = getComplexValue(e2);
+
+    auto isFloatGreaterThan = [](APFloat A, APFloat B) -> llvm::Optional<bool> {
+      // APFloat::compare treats -0.0 == 0.0, but IEEE 754-2018 demands -0 < +0.
+      if (A.isZero() && B.isZero() && (A.isNegative() != B.isNegative()))
+        return !A.isNegative();
+
+      auto cmpRes = A.compare(B);
+      if (cmpRes == APFloat::cmpGreaterThan) return true;
+      if (cmpRes == APFloat::cmpLessThan) return false;
+      if (cmpRes == APFloat::cmpUnordered) return A.isNaN();
+      return llvm::NoneType();
+    };
+
+    // Lexicographic comparison starts with real parts.
+    auto cmpReal =
+        isFloatGreaterThan(leftComplexValue.real(), rightComplexValue.real());
+    if (cmpReal) return *cmpReal ? e1 : e2;
+
+    // Real parts are equal, compare the imag parts.
+    auto cmpImag =
+        isFloatGreaterThan(leftComplexValue.imag(), rightComplexValue.imag());
+    if (cmpImag) return *cmpImag ? e1 : e2;
+
+    // Both real and imag parts are equal, return any (say the left operand).
+    return e1;
```

Comment:
Since we're modelling lexicographic comparison of complex numbers after JAX, let's use [their implementation](https://github.com/google/jax/blob/7c85ca38f45e3ae0d8a80b2d33cf908de9c6564f/jax/_src/lax/lax.py#L2230-L2238). E.g. for `max` that would be:

```
auto pickLhs = lhs.real() == rhs.real() ? lhs.imag() < rhs.imag() : lhs.real() < rhs.real();
return pickLhs ? lhs : rhs.
```

Reply:
The reason we have extra code that that you have suggested is for the following reasons: 

1. The `==` operator of APFloat used `APFloat::compare` which treats `-0 == 0`. OTOH, the spec demands -0< +0.
2. The other part of the code is for propagating the NaNs. However as it is not clear how this should formulated for complex types, we can avoid that code.  At the moment I am outputting the value for which we see a NaN in its real or imag part.

Reply:
1. Let's write a helper function that we'll share between `min` and `max` that does what JAX does to compute `lhs.real() == rhs.real()`. It would seem that in this case that means LLVM's `fcmp oeq` (see [elemental_ir_emitter.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/elemental_ir_emitter.cc#L904-L905)), which is unfortunate because APFloat's `==` seems to be `ueq` (so we'll have to implement our own).

2. "However as it is not clear how this should formulated for complex types, we can avoid that code". Yeah, for the draft spec, let's just do whatever JAX does.